Build break on (at least) Mac OS X 10.12 Sierra

I had a try at building this on OS X Sierra. Since I was unable to find any screenshots, video, releases or status documentation since 2015, and wanted to find out what the project looks like today.

There were 4 problems so far. I have solutions for three of them.
## Problem 1: [configure] `Qt5Core` package configuration not found

```
Call Stack (most recent call first):
  CMakeLists.txt:88 (doxygen_configure)



cpp library
    libopenage                                          [212 sources] [15 generated]
CMake Error at libopenage/CMakeLists.txt:90 (find_package):
  By not providing "FindQt5Core.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "Qt5Core", but
  CMake did not find one.

  Could not find a package configuration file provided by "Qt5Core"
  (requested version 5.4) with any of the following names:

    Qt5CoreConfig.cmake
    qt5core-config.cmake

  Add the installation prefix of "Qt5Core" to CMAKE_PREFIX_PATH or set
  "Qt5Core_DIR" to a directory containing one of the above files.  If
  "Qt5Core" provides a separate development package or SDK, be sure it has
  been installed.
```

Qt5 provides in its `lib/cmake` directory (i.e. `/usr/local/Cellar/qt5/5.6.2/lib/cmake`) some `cmake` scripts (e.g. `Qt5CoreConfig.cmake`). Our `cmake` needs to find these, since they tell it how to link against Qt5's modules (e.g. `Qt5::Core`).
#### Solution

Before running `./configure`, I set the following variable in my shell:

``` bash
CMAKE_PREFIX_PATH="$CMAKE_PREFIX_PATH:/usr/local/Cellar/qt5/5.6.2/lib/cmake"
```
#### Another solution

There is an alternative solution I found that for some reason works.

That is: add `/usr/local/opt/qt5/bin` to your `$PATH`.

I think I remember that this turned out to be only a partial solution. However it's possible my memory is failing me, and that the "remaining problem" with it was simply Problem 3 (which is a problem we encounter with either of the solutions).
## Problem 2: [make] `GL/gl.h` not found

```
➜  openage git:(master) 🍔 make
…
[  1%] Building CXX object libopenage/CMakeFiles/libopenage.dir/engine.cpp.o
In file included from /Users/birch/git/openage/libopenage/engine.cpp:17:
In file included from /Users/birch/git/openage/libopenage/gui_basic.h:12:
/Users/birch/git/openage/libopenage/gui/guisys/public/gui_renderer.h:7:10: fatal error: 'GL/gl.h' file not found
#include <GL/gl.h>
         ^
1 error generated.
```

Apparently the OpenGL headers on Mac live under `OpenGL/gl.h`.
#### Solution

I helped it find the headers by changing `openage/libopenage/gui/guisys/public/gui_renderer.h`:

``` diff
+#ifndef __APPLE__
 #include <GL/gl.h>
+#else
+#include <OpenGL/gl.h>
+#endif
```
## Problem 3 [make]: `QCocoaNativeContext` not found

```
[ 34%] Building CXX object libopenage/CMakeFiles/libopenage.dir/gui/guisys/private/platforms/context_extraction_cocoa.mm.o
/Users/birch/git/openage/libopenage/gui/guisys/private/platforms/context_extraction_cocoa.mm:7:10: fatal error: 
      'QtPlatformHeaders/QCocoaNativeContext' file not found
#include <QtPlatformHeaders/QCocoaNativeContext>
         ^
1 error generated.
```

So, it needs more includes. I used `make VERBOSE=1` to check what includes we were already recruiting:

``` bash
[ 34%] Building CXX object libopenage/CMakeFiles/libopenage.dir/gui/guisys/private/platforms/context_extraction_cocoa.mm.o
cd /Users/birch/git/openage/.bin/clang++-3-5-debug-Oauto-sanitize-none/libopenage && /usr/local/bin/clang++-3.5   -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_QML_LIB -DQT_QUICK_LIB -Dlibopenage_EXPORTS -I/Library/Frameworks/Mono.framework/Headers -I/usr/local/include/freetype2 -I/usr/local/include -I/usr/local/include/opus -I/Users/birch/Library/Frameworks/SDL2.framework/Headers -I/Library/Frameworks/SDL2_image.framework/Headers -iframework /usr/local/Cellar/qt5/5.6.2/lib -isystem /usr/local/Cellar/qt5/5.6.2/lib/QtCore.framework/Headers -isystem /usr/local/Cellar/qt5/5.6.2/./mkspecs/macx-clang -isystem /usr/local/Cellar/qt5/5.6.2/lib/QtQuick.framework/Headers -isystem /usr/local/Cellar/qt5/5.6.2/lib/QtGui.framework/Headers -isystem /System/Library/Frameworks/OpenGL.framework/Headers -isystem /usr/local/Cellar/qt5/5.6.2/lib/QtQml.framework/Headers -isystem /usr/local/Cellar/qt5/5.6.2/lib/QtNetwork.framework/Headers  -Wall -Wextra -pedantic -Wno-mismatched-tags -stdlib=libc++ -std=c++14 -g -O1 -fPIC   -F/Users/birch/Library/Frameworks  -fPIC -std=gnu++14 -o CMakeFiles/libopenage.dir/gui/guisys/private/platforms/context_extraction_cocoa.mm.o -c /Users/birch/git/openage/libopenage/gui/guisys/private/platforms/context_extraction_cocoa.mm
```

So, we already have quite a few Qt header directories included. Such as the `Qt5::Core` module's headers, in the directory: `/usr/local/Cellar/qt5/5.6.2/lib/QtCore.framework/Headers`.

I noticed that Qt provide _no_ `QtPlatformHeaders.framework` module. So we will have to use a different strategy here.
#### Solution:

I added to the `libopenage`'s `CMakeLists.txt`

``` diff
+if(APPLE)
+   set(QTPLATFORM_INCLUDE_DIRS "/usr/local/Cellar/qt5/5.6.2/include")
+endif()

# directories for header inclusion
include_directories(
    ${OPENGL_INCLUDE_DIR}
    ${FREETYPE_INCLUDE_DIRS}
    ${EPOXY_INCLUDE_DIRS}
    ${OPUS_INCLUDE_DIRS}
    ${SDL2_INCLUDE_DIR}
    ${SDL2IMAGE_INCLUDE_DIRS}
    ${HarfBuzz_INCLUDE_DIRS}
+   ${QTPLATFORM_INCLUDE_DIRS}
)
```

Sadly this is hardcoding. Maybe there is a better way. It certainly had the desired effect: `clang` was able to find the `QtPlatformHeaders/QCocoaNativeContext` headers.
### Problem 4 [make]: clang gets very confused trying to read OS X `Foundation`'s `NSObject` header

I am not sure what's going on here.

```
[ 34%] Building CXX object libopenage/CMakeFiles/libopenage.dir/gui/guisys/private/platforms/context_extraction_cocoa.mm.o
In file included from /Users/birch/git/openage/libopenage/gui/guisys/private/platforms/context_extraction_cocoa.mm:7:
In file included from /usr/local/Cellar/qt5/5.6.2/include/QtPlatformHeaders/QCocoaNativeContext:1:
In file included from /usr/local/Cellar/qt5/5.6.2/include/QtPlatformHeaders/qcocoanativecontext.h:38:
In file included from /System/Library/Frameworks/AppKit.framework/Headers/NSOpenGL.h:9:
In file included from /System/Library/Frameworks/Foundation.framework/Headers/Foundation.h:10:
In file included from /System/Library/Frameworks/Foundation.framework/Headers/NSArray.h:5:
/System/Library/Frameworks/Foundation.framework/Headers/NSObject.h:19:21: error: expected a type
- (id)copyWithZone:(nullable NSZone *)zone;
                    ^
/System/Library/Frameworks/Foundation.framework/Headers/NSObject.h:25:28: error: expected a type
- (id)mutableCopyWithZone:(nullable NSZone *)zone;
                           ^
/System/Library/Frameworks/Foundation.framework/Headers/NSObject.h:32:4: error: expected a type
- (nullable instancetype)initWithCoder:(NSCoder *)aDecoder; // NS_DESIGNATED_INITIALIZER
   ^
/System/Library/Frameworks/Foundation.framework/Headers/NSObject.h:44:12: error: unknown property attribute 'class'
@property (class, readonly) BOOL supportsSecureCoding;
           ^
/System/Library/Frameworks/Foundation.framework/Headers/NSObject.h:57:4: error: expected a type
- (nullable id)replacementObjectForCoder:(NSCoder *)aCoder;
   ^
/System/Library/Frameworks/Foundation.framework/Headers/NSObject.h:58:4: error: expected a type
- (nullable id)awakeAfterUsingCoder:(NSCoder *)aDecoder NS_REPLACES_RECEIVER;
   ^
In file included from /Users/birch/git/openage/libopenage/gui/guisys/private/platforms/context_extraction_cocoa.mm:7:
In file included from /usr/local/Cellar/qt5/5.6.2/include/QtPlatformHeaders/QCocoaNativeContext:1:
In file included from /usr/local/Cellar/qt5/5.6.2/include/QtPlatformHeaders/qcocoanativecontext.h:38:
In file included from /System/Library/Frameworks/AppKit.framework/Headers/NSOpenGL.h:9:
In file included from /System/Library/Frameworks/Foundation.framework/Headers/Foundation.h:10:
In file included from /System/Library/Frameworks/Foundation.framework/Headers/NSArray.h:6:
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:7:15: error: expected ';' after @class
@class NSArray<ObjectType>;
              ^
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:7:16: error: cannot find protocol declaration for 'ObjectType'
@class NSArray<ObjectType>;
               ^
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:34:25: error: cannot find protocol declaration for 'ObjectType'
@interface NSEnumerator<ObjectType> : NSObject <NSFastEnumeration>
                        ^
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:34:37: error: expected unqualified-id
@interface NSEnumerator<ObjectType> : NSObject <NSFastEnumeration>
                                    ^
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:36:4: error: expected a type
- (nullable ObjectType)nextObject;
   ^
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:40:25: error: cannot find protocol declaration for 'ObjectType'
@interface NSEnumerator<ObjectType> (NSExtendedEnumerator)
                        ^
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:40:1: error: duplicate interface definition for class 'NSEnumerator'
@interface NSEnumerator<ObjectType> (NSExtendedEnumerator)
^
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:34:12: note: previous definition is here
@interface NSEnumerator<ObjectType> : NSObject <NSFastEnumeration>
           ^
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:40:37: error: method type specifier must start with '-' or '+'
@interface NSEnumerator<ObjectType> (NSExtendedEnumerator)
                                    ^
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:40:38: error: expected a type
@interface NSEnumerator<ObjectType> (NSExtendedEnumerator)
                                     ^
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:42:1: error: expected selector for Objective-C method
@property (readonly, copy) NSArray<ObjectType> *allObjects;
^
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:42:28: error: unknown type name 'NSArray'
@property (readonly, copy) NSArray<ObjectType> *allObjects;
                           ^
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:42:35: error: expected member name or ';' after declaration specifiers
@property (readonly, copy) NSArray<ObjectType> *allObjects;
                           ~~~~~~~^
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:42:1: error: property requires fields to be named
@property (readonly, copy) NSArray<ObjectType> *allObjects;
^                          ~~~~~~~
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
```

I think it doesn't understand the syntax of these headers (Foundation is a system framework that comes with OS X). So maybe it was expecting C++, but encountered Objective-C, and exploded? But at the same time: it is aware of Objective-C, as evidenced by the message:

> expected selector for Objective-C method

So, I don't really get what happened. Anyway, I thought I may as well report my findings.

```
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:7:15: error: expected ';' after @class
@class NSArray<ObjectType>;
```

That particular error suggests to me that the compiler has never heard of generic types. So certainly I suspect it's encountering an unexpected programming language.

--- a/copying.md
+++ b/copying.md
@@ -89,6 +89,7 @@ _the openage authors_ are:
 | Johannes Walcher            | tomatower                   | johannes.walcher@stusta.de            |
 | Akritas Akritidis           | MaanooAk                    | akritasak@gmail.com                   |
 | Edgard Mota                 | edgardmota                  | edgardmota@gmail.com                  |
+| Boris Dušek                 | dusek                       | me@dusek.me                           |
 
 If you're a first-time commiter, add yourself to the above list. This is not
 just for legal reasons, but also to keep an overview of all those nicknames.

--- a/libopenage/CMakeLists.txt
+++ b/libopenage/CMakeLists.txt
@@ -68,6 +68,11 @@ find_path(FREETYPE_INCLUDE_DIRS freetype/freetype.h HINTS /usr/include/freetype2
 include(FindPkgConfig)
 include(FindPackageHandleStandardArgs)
 
+# provide apple qt5 location
+if(APPLE)
+	list(APPEND CMAKE_PREFIX_PATH /usr/local/opt/qt5)
+endif()
+
 # windows does not have libm
 if(NOT WIN32)
 	find_library(MATH_LIB m)
@@ -139,6 +144,10 @@ get_config_option_string()
 configure_file(config.h.in ${CMAKE_CURRENT_SOURCE_DIR}/config.h)
 configure_file(config.cpp.in ${CMAKE_CURRENT_SOURCE_DIR}/config.cpp)
 
+if(APPLE)
+	find_path(QTPLATFORM_INCLUDE_DIRS QtPlatformHeaders "/usr/local/opt/qt5/include")
+endif()
+
 # directories for header inclusion
 include_directories(
 	${OPENGL_INCLUDE_DIR}
@@ -148,6 +157,7 @@ include_directories(
 	${SDL2_INCLUDE_DIR}
 	${SDL2IMAGE_INCLUDE_DIRS}
 	${HarfBuzz_INCLUDE_DIRS}
+        ${QTPLATFORM_INCLUDE_DIRS}
 )
 
 # link the executable to those libraries

--- a/libopenage/gui/guisys/public/gui_renderer.h
+++ b/libopenage/gui/guisys/public/gui_renderer.h
@@ -4,7 +4,11 @@
 
 #include <memory>
 
+#ifndef __APPLE__
 #include <GL/gl.h>
+#else
+#include <OpenGL/gl.h>
+#endif
 
 struct SDL_Window;
 

